nvd/nvd.go

```diff
@@ -1,162 +1,250 @@
+		return 0, xerrors.Errorf("unable to save entry: %w", err)
+	}
+	return entry.TotalResults, nil
+}
 
-		if err = utils.SaveCVEPerYear(filepath.Join(utils.VulnListDir(), feedDir), cveID, item); err != nil {
-			return xerrors.Errorf("failed to save NVD CVE detail: %w", err)
+func save(entry Entry, lastModEndDate string) error {
+	for _, cve := range entry.Vulnerabilities {
+		if err := utils.SaveCVEPerYear(filepath.Join(utils.VulnListDir(), apiDir), cve.Cve.ID, cve); err != nil {
+			return xerrors.Errorf("unable to write %s: %w", cve.Cve.ID, err)
 		}
 	}
+	// we need to save LastModEndDate to avoid saving wrong time if we get errors when saving all CVEs (first run)
+	// after this we can switch to entry.Timestamp
+	timestamp, err := time.Parse(nvdTimeFormat, lastModEndDate)
+	if err != nil {
+		return xerrors.Errorf("unable to parse timestamp: %w", err)
+	}
+	// update the Last_updated.json file after saving each entry
+	// to avoid overwriting this entry if we fail to save the next entry
+	err = utils.SetLastUpdatedDate("nvd", timestamp)
+	if err != nil {
+		return xerrors.Errorf("unable to save last_updated.json file: %w", err)
+	}
 	return nil
 }
 
-func fetchLastModifiedDate(feed string) (time.Time, error) {
-	log.Printf("Fetching NVD metadata(%s)...\n", feed)
-
-	url := fmt.Sprintf("%s/nvdcve-1.1-%s.meta", baseURL, feed)
-	res, err := utils.FetchURL(url, "", 5)
+func (u Updater) getEntry(url string) (Entry, error) {
+	var entry Entry
+	r, err := fetchURL(url, u.apiKey, u.retry)
 	if err != nil {
-		return time.Time{}, xerrors.Errorf("fetch error: %w", err)
+		return entry, xerrors.Errorf("unable to fetch: %w", err)
+	}
+	if r != nil {
+		if err = json.NewDecoder(r).Decode(&entry); err != nil {
+			return entry, xerrors.Errorf("unable to decode response for %q: %w", url, err)
+		}
+		return entry, nil
 	}
 
-	scanner := bufio.NewScanner(bytes.NewBuffer(res))
-	for scanner.Scan() {
-		line := scanner.Text()
-		s := strings.SplitN(line, ":", 2)
-		if len(s) != 2 {
+	return entry, xerrors.Errorf("unable to get entry from %q", url)
+}
+
+func fetchURL(url, apiKey string, retry int) (io.ReadCloser, error) {
+	var c http.Client
+	for i := 0; i <= retry; i++ {
+		req, err := http.NewRequest("GET", url, nil)
+		if err != nil {
+			return nil, xerrors.Errorf("unable to build request for %q: %w", url, err)
+		}
+		if apiKey != "" {
+			req.Header.Set("apiKey", apiKey)
+		}
+
+		resp, err := c.Do(req)
+		if err != nil {
+			log.Printf("Response error: %s. Try to get Entry again", err)
 			continue
 		}
-		if s[0] == "lastModifiedDate" {
-			t, err := time.Parse(time.RFC3339, s[1])
-			if err != nil {
-				return time.Time{}, err
-			}
-			return t, nil
+		switch resp.StatusCode {
+		case http.StatusForbidden:
+			log.Printf("NVD rate limit. Waiting to gain access")
+			// NVD limits:
+			// Without API key: 5 requests / 30 seconds window
+			// With API key: 50 requests / 30 seconds window
+			time.Sleep(30 * time.Second)
```

Comment:
By the way, there is a package `https://pkg.go.dev/golang.org/x/time/rate#NewLimiter` that provides Limiter.

Example of use:
```go
rate.NewLimiter(rate.Every(30*time.Second), 50) // with key
```